Add Readme for Restaurant Demo #313 Type Documentation

mhosankalp opened the pull request:
## Description
Add Readme for Restaurant Demo #313
https://github.com/OpenMined/PyDP/issues/313
Type: Documentation :books: 

## Affected Dependencies
No Dependencies. The readme file is independent. Along with this readme, there are 4 images of graph which are referencing in the Readme file

## How has this been tested?
This has been tested in my repository. No Integration testing needed, as this is a documentation for the Restaurant examples.

## Checklist
- I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- My changes are covered by tests

mhosankalp committed:
Create s1.txt

mhosankalp committed:
Add files via upload

mhosankalp committed:
Create README.md

mhosankalp committed:
Delete s1.txt

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,260 @@
+# Example using the Differential Privacy Library PyDP
+
+## Restaurant
```
could you change this to "Restaurant Demo"

mhosankalp replied:
This is resolved. I have updated "Restaurant Demo" in the header

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,260 @@
+# Example using the Differential Privacy Library PyDP
+
+## Restaurant
+
```
pls add google's dp java example link as ref. 
You may also look at the intro from here as a reference #311

mhosankalp replied:
I have added Google Java as well as our notebook link for reference

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,260 @@
+business statistics with her visitors. Alice knows when visitors enter the
+restaurant and how much time and money they spend there. To ensure that
+visitors' privacy is preserved, Alice decides to use the Differential Privacy
+library.
```
pls mention PyDP here

mhosankalp replied:
Done

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,260 @@
+* Sum-up revenue per day of the week with preaggregation
+
+Non Private Count is the raw counts and output
+Private count is anonymized Counts and output using Differential Privacy library
```
PyDP lib

mhosankalp replied:
Done

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,260 @@
+from the actual ones but the overall trend is preserved. For example, you can
+clearly see that the restaurant is more busy during lunch and dinner time.
+
+![Daily counts](img/counts_per_hour.png)
```
Can't see the image. Pls, attach the correct link. -- this seems to be the case for all the images in this readme.

mhosankalp replied:
I am able to see the images

shaistha24 commented on examples/restaurant_demo/README.md:
````diff
@@ -0,0 +1,260 @@
+uses `Count` to calculate the differentially private count of visits for a
+single day.
+
+```java
````
this is the code snippet from google's example. pls add the one from pydp lib example

mhosankalp replied:
Done

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,260 @@
+
+Navigate to `PyDP/examples/restaurant_demo` folder. Execute `Python restaurant.py`
+
+The output will display Private and Non-Private counts for:
```
mainly 3 scenarios Count visits by hour of day,  Count visits by day of week,  Sum-up revenue per day of the week

Count visits by day of week -- has two cases - max_contribution = 7 or 3 days/week
Sum-up revenue per day of the week -- same but with upper bound set to 65  --  max_contribution= 4 days/week

mhosankalp replied:
Corrected  to include only three scenario:
 1) Count visits by hour of day
 2) Count visits by day of week
 3) Sum-up revenue per day of the week

shaistha24 reviewed (changes requested):
Apart from the suggested changes, you may as well want to change the 'we" with "you" as we should strive to make the language accessible and inclusive. Pls refer to this https://github.com/OpenMined/PyDP/pull/307#pullrequestreview-508552678 for more details. Thanks 



you may also refer to the jupyter notebook in #311 for refernce if needed. ( Note: only the necessary exaplantion has been included in the notebook)

#

mhosankalp committed:
Update README.md

shaistha24 reviewed (changes requested):
@mhosankalp Pls refer to all of my previous review comments and make the necessary changes.  The images or graphs are still not visible. Looks like you have misunderstood my reviews I did not ask you to remove any of the explanation parts but instead, tone is as per the google style guide using "you" in place of "we" - i.e., writing in second person, making it easier for everyone to understand. 

Also, making "Restaurant Demo" as the main heading would be better -  followed by adding the link to googles github readme link 

Pls use this notebook as a reference   https://github.com/OpenMined/PyDP/blob/dev/examples/restaurant_demo/restaurant_demo.ipynb

shaistha24 commented:
I apologize for the late review but pls @me so that I might recieve  github notification. Sometimes I don't receive any notification and won't know about it unless i check it myself. Apologies.

mhosankalp committed:
Update README.md

mhosankalp committed:
Update README.md

mhosankalp committed:
Update README.md

mhosankalp committed:
Update README.md

mhosankalp commented:
@shaistha24, I have completed all the updates. Please let me know if any further changes. The images I am not sure why you are not able to see as I am able to see it

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,259 @@
+# Example using the Differential Privacy Library PyDP
```
Pls remove this

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,259 @@
+# Example using the Differential Privacy Library PyDP
+
+## Restaurant Demo
+
```
Pls add something like this down: (pls ignore the part you have already added and pls add in the part which was not present)

PyDP library version for Google's Java Differential Privacy Library example - https://github.com/google/differential-privacy/tree/master/examples/java.

Imagine a fictional restaurant owner named Alice who would like to share business statistics with her visitors or potential clients. Alice knows when visitors enter the restaurant and how much time and money they spend there. To ensure that visitor's privacy is preserved, Alice decides to use the Differential Privacy library in this case PyDP library.

Alice wants to share the information with potential clients which include 3 main scenarios in total.

    Count visits by an hour of the day: Count how many visitors enter the restaurant at every hour of a particular day.
    Count visits by day of the week: Count how many visitors enter the restaurant each day in a week. This in itself includes Two cases- 
    Sum-up revenue per day of the week: Calculate the sum of the restaurant revenue per weekday.

Both the Non-private and the Private values for the above cases are calculated down below along with the graphical representation of the results and their comparisons.

Notebook Implementation for the same can be found [here]( https://github.com/OpenMined/PyDP/blob/dev/examples/restaurant_demo/restaurant_demo.ipynb)

shaistha24 commented on examples/restaurant_demo/README.md:
````diff
@@ -0,0 +1,259 @@
+                MAX_EUROS_SPENT_2,
+                l0_sensitivity=SUM_MAX_CONTRIBUTED_DAYS,
+            )
+```
````
Everything from the Java example has been implemented - like I have mentioned before the notebook has only quick notable explanation. Pls add all the remaining part from the googles readme with PyDP code such as 

Clamping individual contributions and below 
https://github.com/OpenMined/PyDP/blob/dev/examples/restaurant_demo/restaurant.py#L89 code from pydp

shaistha24 replied:
the explanation part

shaistha24 commented on examples/restaurant_demo/README.md:
````diff
@@ -0,0 +1,259 @@
+the visitors' spendings for a single day.
+
+```
+# Cap the maximum number of visiting days at 4 per each visitor (any number above will not be taken into account)
````
Sorry for not noticing it before pls check the code ... this might confuse the readers

https://github.com/OpenMined/PyDP/blob/91cfbeacd3445646620e9494c609040c42f31efa/examples/restaurant_demo/restaurant.py#L220

and https://github.com/OpenMined/PyDP/blob/91cfbeacd3445646620e9494c609040c42f31efa/examples/restaurant_demo/restaurant.py#L254

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,259 @@
+
+## Partitions and contributions
+
+We say that the resulting aggregated data is split into *partitions*. The bar
```
Pls change the "We" to "you" that is the second person or change the sentence.  for example - the resulting aggregated data is split into *partitions*.  ... 

Pls refer to this review https://github.com/OpenMined/PyDP/pull/307#pullrequestreview-508552678 for more detail

Such as 
    In your local (virtual) env, you can install the TensorFlow team's notebook tools:

      python3 -m pip install -U git+https://github.com/tensorflow/docs

 Then use the nblint linter with the --styles=google flag:

         python3 -m tensorflow_docs.tools.nblint --styles=google your_notebook.ipynb

This doesn't highlight everything, but it's good. Check out this README for more options.

shaistha24 reviewed (changes requested):
Nice work @mhosankalp, pls make the requested changes and this should be good to go.

mhosankalp committed:
Update README.md

mhosankalp committed:
Fixing the broken link

an introduction to PyDP and the carrots demo link were broken. It was fixed here.

mhosankalp committed:
fixing broken link

fixing broken link for Error Handling in PyDP

mhosankalp committed:
Merge pull request #1 from mhosankalp/mhosankalp-patch-1-Issue336

fixing broken link

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+visitors' privacy is preserved, Alice decides to use the Differential Privacy
+library in this case PyDP library.
+
+Alice wants to share the information with potential clients which include 3 main scenarios in total.
```
you may just change it to " includes 4 scenarios:"

shaistha24 commented on examples/restaurant_demo/README.md:
````diff
@@ -0,0 +1,279 @@
+Alice wants to share the information with potential clients which include 3 main scenarios in total.
+
+```
+Count visits by an hour of the day: Count how many visitors enter the restaurant at every hour of a particular day.
````
Hi, Looks good but for better readability, I would like to suggest if you could uncomment them i.e, remove ``` and write the scenarios in bold eg **Count visits by an Hour of the day:**

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+## Restaurant Demo
+
+PyDP library version for Google's Java Differential Privacy Library example - 
```
Thanks. Could you pls create a link (inline style) for any other links added **throughout the readme**  for example in this case - line 3

`PyDP library version for [Google's Java Differential Privacy Library](https://github.com/google/differential-privacy/tree/master/examples/java)example`

Markdown Cheatsheet: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet

shaistha24 commented on examples/restaurant_demo/README.md:
````diff
@@ -0,0 +1,279 @@
+Sum-up revenue per day of the week with preaggregation
+```
+    
+Both the Non-private and the Private values for the above cases are calculated down below along with the graphical representation of the results and their comparisons.
````
You may remove this line  - as we are not really showing the results here, right! It's just a quick introductory material.

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+
+Notebook Implementation for the same can be found https://github.com/OpenMined/PyDP/blob/dev/examples/Tutorial_2-restaurant_demo/restaurant_demo.ipynb
+
+## Installation
```
Nice! For a clear understanding, I suggest the title could be changed to "To Run the Demo" instead of installation. 

Also ` pip install python-dp`  as codefor better code readability.

thanks

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+* Sum-up revenue per day of the week
+* Sum-up revenue per day of the week with preaggregation
+
+Non Private Count is the raw counts and output
```
Could you pls rephrase it like this:

"Non-Private Count is the raw count whereas the Private Count is anonymized Count generated using PyDP library."

As it has already been mentioned that PyDP is the DP lib in use, we can directly write it as PyDP to avoid confusions.

shaistha24 replied:
" has been used to quote the requested change. Pls do not add it to the readme by mistake. Thank you.

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+restaurant.
+
+
+This triggers the logic of `CountVisitsPerHour`. It reads the daily statistics
```
could you pls **REMOVE line 55- 63**. It will be confusing for the readers as this part is from the NoteBook and the function name in .py is different. Explanation after the bar plot is enough to understand the concept.

Pls **do not remove** line 65 and below stay i.e. the bar plot part.

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+Why is this important? Differential Privacy adjusts the amount of noise to mask
+contributions of each visitor. More contributions require more noise.
+
+Next, analyse for the remaining three cases:
```
Pls REMOVE line 108 - 112, you don't necessarily have to keep reminding the reader what's up next. We have already mentioned the scenarios above.

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+Differential Privacy adds some amount of random noise to hide contributions of
+an individual. The more contributions an individual has, the larger the noise
+is. This affects the utility of the data. In order to preserve the data utility,
+you have to made an approximate estimate of how many times a week a person may visit a
```
line 168 - " you have to **make** an approximate estimate of how many times a week a person may..."

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+restaurant on average, and assumed that the value is around 3 instead of scaling
+the noise by the factor of 7.
+
+You have also pre-processed the input data and discarded all exceeding visits. The
```
Pls change Line 172 for better readability to 

" The input data can be pre-processed discarding all the exceeding visits using `bound_visits_per_week(self._day_visits, COUNT_MAX_CONTRIBUTED_DAYS)`. "

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+to keep in mind that the library allows you to specify maximum amount of
+contributions, but doesn't validate that it is respected.
+
+Next, analyse for the remaining three cases:
```
Pls remove line 177 - 180

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+The previous example demonstrates how the contributed partitions are bounded.
+Now, you will demonstrate how individual contributions are clamped. Imagine Alice
+decides to calculate the sum of the restaurant revenue per week day in a
+differentially private way. For this she needs to sum up the visitors’ daily
```
line 187: pls change it to "differentially private way. For this, she needs to sum up the **visitor's** daily"

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+
+![Daily sums](img/sums_per_day.png)
+
+The code below uses `BoundedSum` to calculate the differentially private sums of
```
Pls change line 197-198 to 

" `BoundedSum` is **used to** calculate the differentially private sums of the **visitor's** spendings for a single day."

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -0,0 +1,279 @@
+the input values are clamped more aggressively, which can decrease utility as
+well.
+
+Let's set the `upper` bound of`BoundedSum` to 65 to reflect the approximate
```
Line 253 - could you pls change to 

"The `upper` bound of`BoundedSum` to 65 to reflect the...." 

thank you

shaistha24 reviewed (changes requested):
Hi, @mhosankalp  Nice work! Thank you for being patient and making requested changes. I have commented a few minor edits to be done some minor spelling mistakes or sentence phrasing else it's all good. 

Should be quick and easy, as I have commented on the edits. For most of it, you will just have to copy-paste it.

Also, could you pls update your Repo as the file structure and naming have been updated. 

Looking forward to seeing your PR finally getting merged after this.

mhosankalp committed:
update Changes raised on 01/22/2021

update Changes raised on 01/22/2021

mhosankalp committed:
update Changes raised on 01/22/2021

update Changes raised on 01/22/2021

mhosankalp committed:
Update README.md

mhosankalp committed:
Update README.md

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -1,32 +1,28 @@
 
-PyDP library version for Google's Java Differential Privacy Library example - 
-https://github.com/google/differential-privacy/tree/master/examples/java.
+PyDP library version for Google's Java Differential Privacy Library example can be found [here](https://github.com/google/differential-privacy/tree/master/examples/java).
```
Hi, this kinda sounds confusing. 

Could you pls change this to, 

PyDP library version for [Google's Java Differential Privacy Library - Restaurant Example](https://github.com/google/differential-privacy/tree/master/examples/java).

anchor link would be less confusing and better. Please change it to the above.

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -1,32 +1,28 @@
-Both the Non-private and the Private values for the above cases are calculated down below along with the graphical representation of the results and their comparisons.
 
-Notebook Implementation for the same can be found https://github.com/OpenMined/PyDP/blob/dev/examples/Tutorial_2-restaurant_demo/restaurant_demo.ipynb
+#### 1) Count visits by an hour of the day: Count how many visitors enter the restaurant at every hour of a particular day.
```
could you pls put only " Count visits by an hour of the day:", "Count visits by day of the week:" etc bold leaving the description in normal. 

to do so you may put the text  in between double * i.e., ** ** 

ref: https://www.ionos.com/digitalguide/websites/web-development/markdown/

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -1,32 +1,28 @@
 
-## Installation
+## To Run the Demo
 To install PyDP, use the PyPI package manager:
```
Could you pls remove "to"
"Install PyDP, using the PyPI package manager:"

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -1,32 +1,28 @@
 
-pip install python-dp
-(If you have pip3 separately for Python 3.x, use pip3 install python-dp.)
+`pip install python-dp` (If you have pip3 separately for Python 3.x, use pip3 install python-dp.)
```
Pls add the below two-line (apologies I should have noticed this before!) 

or 
`pip3 install python-dp`

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -1,32 +1,28 @@
-(If you have pip3 separately for Python 3.x, use pip3 install python-dp.)
+`pip install python-dp` (If you have pip3 separately for Python 3.x, use pip3 install python-dp.)
 
 Navigate to `PyDP/examples/restaurant_demo` folder. Execute `Python restaurant.py`
```
Pls add "and"

Navigate to `PyDP/examples/restaurant_demo` folder and execute `Python restaurant.py`

mhosankalp committed:
to sync forked branch with actual branch Merge branch 'dev' of https://github.com/OpenMined/PyDP into dev

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -1,32 +1,28 @@
 
-Notebook Implementation for the same can be found https://github.com/OpenMined/PyDP/blob/dev/examples/Tutorial_2-restaurant_demo/restaurant_demo.ipynb
+#### 1) Count visits by an hour of the day: Count how many visitors enter the restaurant at every hour of a particular day.
+#### 2) Count visits by day of the week: Count how many visitors enter the restaurant each day in a week. This includes two cases:
```
You have not mentioned the two cases involved? pls, add them.

mhosankalp replied:
I do not have two cases in explanation. So I removed the text "This includes two cases:"

mhosankalp committed:
syncing with original repo for restaurant demo

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -165,26 +143,18 @@ partitions a visitor may contribute to. You might notice that the value of
 Now, you will demonstrate how individual contributions are clamped. Imagine Alice
 decides to calculate the sum of the restaurant revenue per week day in a
-differentially private way. For this she needs to sum up the visitors’ daily
+differentially private way. For this, she needs to sum up the <b>visitor's</b> daily
```
could you pls unbold the "visitor's"?

shaistha24 commented on examples/restaurant_demo/README.md:
```diff
@@ -194,8 +164,7 @@ The results are illustrated in the image below.
 
-The code below uses `BoundedSum` to calculate the differentially private sums of
-the visitors' spendings for a single day.
+`BoundedSum` is used to calculate the differentially private sums of the <b>visitor's</b> spendings for a single day.
```
pls unbold "visitor's"

shaistha24 reviewed (changes requested):
Thanks, @mhosankalp  I did see a few errors like bolding "visitor's"  etc pls check it and it should be good to go.

mhosankalp committed:
changes as per review comment on 1/25/2021

changes as commented on 1/25/2021

mhosankalp committed:
changes as review comment on 1/25/2021

changes as review comment on 1/25/2021

mhosankalp committed:
Update README.md

mhosankalp commented:
Thanks @shaistha24 for the review. I was able to resolve all of them. Also my local repo has been updated to the main. Let me know if you see any other issue.

Thanks and Regards,
Sankalp

shaistha24 reviewed (approved):
Woot!! Nice work @mhosankalp! Thank you for resolving all the review comments. Hope to see amazing write-ups from you in the future too!

chinmayshah99 commented:
Amazing work @mhosankalp! Looking forward to your next PR.
